tsdb/head.go

```diff
@@ -1667,26 +1667,34 @@ func (h *Head) mmapHeadChunks() {
 	var count int
 	for i := 0; i < h.series.size; i++ {
 		h.series.locks[i].RLock()
-		for _, all := range h.series.hashes[i] {
-			for _, series := range all {
-				series.Lock()
-				count += series.mmapChunks(h.chunkDiskMapper)
-				series.Unlock()
-			}
+		for _, series := range h.series.series[i] {
+			series.Lock()
+			count += series.mmapChunks(h.chunkDiskMapper)
+			series.Unlock()
 		}
 		h.series.locks[i].RUnlock()
 	}
 	h.metrics.mmapChunksTotal.Add(float64(count))
 }
 
-// seriesHashmap is a simple hashmap for memSeries by their label set. It is built
-// on top of a regular hashmap and holds a slice of series to resolve hash collisions.
+// seriesHashmap lets TSDB find a memSeries by its label set, via a 64-bit hash.
+// There is one map for the common case where the hash value is unique, and a
+// second map for the case that two series have the same hash value.
+// Each series is in only one of the maps.
 // Its methods require the hash to be submitted with it to avoid re-computations throughout
 // the code.
-type seriesHashmap map[uint64][]*memSeries
+type seriesHashmap struct {
+	unique    map[uint64]*memSeries
+	conflicts map[uint64][]*memSeries
+}
 
-func (m seriesHashmap) get(hash uint64, lset labels.Labels) *memSeries {
-	for _, s := range m[hash] {
+func (m *seriesHashmap) get(hash uint64, lset labels.Labels) *memSeries {
+	if s, found := m.unique[hash]; found {
+		if labels.Equal(s.lset, lset) {
+			return s
+		}
+	}
+	for _, s := range m.conflicts[hash] {
```

Comment:
I'm a bit confused here, perhaps you can help me clarify. 

The first series hash set into the map will be set into `m.unique[hash] = s` by the `set()` method. The second time, if there is a conflict, it will be set into `m.conflicts`. 

So I see that in this method you first check if the hash is in the unique map and then you do a label set comparison. If it doesnt match you look into the conflicts.

I wonder if we shouldnt look first into the conflicts and then into the uniques.

Reply:
Yes, the "happy path" is that the first match is the one you want.  Hash conflicts are expected once in a billion sets of labels.

I think of `unique+conflicts` as one list of series, same as before this PR, so you start at the first one and check that, then move on to the next and so on.

In the case we do have a hash conflict, we can't predict whether the first or second series will occur more often and hence which one we should favour.

Reply:
> I think of unique+conflicts as one list of series, same as before this PR, so you start at the first one and check that, then move on to the next and so on.

> In the case we do have a hash conflict, we can't predict whether the first or second series will occur more often and hence which one we should favour.

Thanks, this makes more sense to me now!